### PR TITLE
Pull Request for the 2nd milestone within the overall Tic Tac Toe project

### PR DESCRIPTION
**Pull request by jose-Abel**
Hello,

This is the Pull Request for the 2nd milestone within the Tic Tac Toe project where the main.rb is responsible for user interactions, where ask for users' names and their choices in the board and then base on the logic of the Tic Tac Toe game determine which user wins or if it's a tie. This implementation of the user-interface milestone so far consists of 13 methods: 

- The main method where it initializes the 2 players hashes, the board hash, and define the while loop that keeps the game going with the conditions of when should break if there's a win or there's a tie.

- The intro method where the initial instructions are displayed at the beginning of the game.

- The choose_your_player method where each player provides it's named to the program to keep track of who are the first and the second players. 

- The name_not_empty method in charge of making sure the name input is not an empty string, if it is then it goes into recursion into at least a one-character string is giving to the name input.

- The repeated_name method in charge of checking if the name the second player is using is already used by the first player and if so it goes into a recursion repeating that same method.

- The paint_canvas method, this method is in charge of displaying the Tic Tac Toe board and user's selection in the screen base on the user mark string of the respective user define in the user hash.

- The get_user_input method in charge of only getting the users' input of their choice on the board.

- The move_in_board method in charge of checking if the player's choice is available on the board as a key and if that key is empty, if it is then it adds the user's mark (X or O depending on the user) to its respective key in the board hash, then it also adds the user's move to the player's moves array. If the player's choice is not available on the board then send a message that is not available and goes into a recursion calling itself and asking the players to please provide a new choice.

- The winning_moves method where checks the win condition base on the player's moves array and base in the Tic Tac Toe game logic. This method was previously just one but had to divide into smaller methods for the complexity and now is in charge of coordinating the launch of the other methods named horizontal_checking, vertical_checking, and diagonal_checking.

- The last method is the hash_has_blank method, this method was refactored because of the changes to the board from an array to a hash. The only purpose of this method is to check if there are any key with an empty value in the board, if there is returns true so that the game continues, if there isn't any key with an empty value, meaning there are no more empty spaces in the board, returns false so main take that false and ends the game.

Hoping to get amazing feedback, and happy coding!

### CODE REVIEW

File: bin/main.rb [hunk 1]
```diff
@@ -1,3 +1,146 @@
...
+  if player[:moves].include?('A1') && player[:moves].include?('B2') && player[:moves].include?('C3')
+    puts "#{player[:name]} you won!"
+    true
+  elsif player[:moves].include?('A3') && player[:moves].include?('B2') && player[:moves].include?('C1')
+    puts "#{player[:name]} you won!"
+    true
+  else
+    false
+  end
+end
```
Comment (bertil291utn) [inline bin/main.rb]: ## Use static variables
- [x] I'd suggest using static variables that already have winning moves, and then use `my_array.any` property to check that if just one array group has all these numbers there's a winner 
Example
```ruby
winning_moves=[[1,2,3],[3,5,7],[1,5,9]...]
```

**Review by bertil291utn — CHANGES_REQUESTED** [review]
# **Status:** Changes requested 🟠 
<p align="center"> image <br><strong>Not yet</strong> </p>

# Hey guys, you've been doing a great job, and also I hope you enjoyed coding.

### Your work is amazing you master it GitHub Flow you have a professional README file. 

However, in order to accomplish all the [requirements](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/), you required to do these changes, listed below:


* [x] `[REQUIRED]` game displays board after the player move
* [x] `[REQUIRED]` game informs player if the selected move is a winning move
* [x] `[REQUIRED]` game informs player if the selected move is a draw move
* [x] `[REQUIRED]` game repeats all actions for the next player's move

### Summary
- Remember guys your code should not work yet, that's for the next milestone, for this one you could just add as inline comments or mocking up the methods. Example
```ruby
def validMove?(player)
  value=false
  value=true if player.move
value
end
```
-  Don't forget to work with inline comments below this comment box.

***

Guys work in the requested changes above and request another review.
Good luck. 

> “Never stop, never conform until what is good becomes even better, and what’s better becomes excellent.”

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Abel Ramirez Frontany** 
Great with your first submission for this project :muscle:  :muscle: 
We still need to fix some details:

- [x] Your game must choose the board whenever the position is updated for a valid user movement.
- [x] You must validation the name of the two users are different
- [x] You must validate the name of the users contains at least one character
- [x] You should favor the division of your code in different functions and files. Create at least one class for the basic functions of your game, save your class in the lib folder.
- [x] Use OOP concepts, your project should have classes (at least one), private methods, and attribute accessors.

Please request another review once you are ready! :+1: 

Happy coding :relaxed:

Oscar Andrés Russi Porras

**Comment by jose-Abel** [comment]
> ### Status: Changes Required ♻️
> **Hello Abel Ramirez Frontany**
> Great with your first submission for this project 💪 💪
> We still need to fix some details:
> 
> * [x]  Your game must choose the board whenever the position is updated for a valid user movement.
> * [x]  You must validation the name of the two users are different
> * [x]  You must validate the name of the users contains at least one character
> * [x]  You should favor the division of your code in different functions and files. Create at least one class for the basic functions of your game, save your class in the lib folder.
> * [x]  Use OOP concepts, your project should have classes (at least one), private methods, and attribute accessors.
> 
> Please request another review once you are ready! 👍
> 
> Happy coding ☺️
> 
> Oscar Andrés Russi Porras

Hello, I'm trying to reach out to you through slack and it hasn't been successful. Among the improvements you mentioned is that I have to make different files and use more OOP concepts creating different classes and attribute. The thing is I know how to do that, but this milestone doesn't require those changes, keep in mind this is the user-interface milestone and I guess that those changes of making it more OOP are necessary for the game-logic milestone. Since this project has to be approved in phases, such that this user-interface milestone has to be approved separated from the game-logic milestone, I think you are requesting changes that are needed to get the next milestone approved. 

I appreciate a lot your feedback and would like to know if it's me that is mistaken. 

Best regards!

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Abel Ramirez Frontany** 
Great job with the improvements since the last review :muscle:  :muscle: 
Your project meets most of the requirements :clap: :clap: :clap:
And you are right, the OOP principles are not a requirement for the current milestone, my apologies. Still, there is one last detail to correct:

- [x] game should displays board after the player move (this is part of the checklist for this milestone).

Please request another review once you are ready! :+1: 

Happy coding :relaxed:

Oscar Andrés Russi Porras

**Review by bertil291utn — APPROVED** [review]
# **Status:** Approved ✅

## Hey guys, you've been doing a great job, and also I hope you enjoyed coding.



<h2 align="center">Ship it </h2>

<p align="center">image</p>

Please merge to develop branch and move on with the next milestone

You're doing  a great job keep going 💻 🚀 🔥

> “Never stop, never conform until what is good becomes even better, and what’s better becomes excellent.”

**Pull request merged**